### PR TITLE
[IMP] mrp: Cache calls to `_bom_find()`

### PR DESCRIPTION
**Pull request by yajo**
Description of the issue/feature this PR addresses:
This method [is called many times from `explode()`][1]. Each time it's called it will return the same result based on the input conditions, and it's very common that they are repeated, so by adding a cache to it we boost a lot the product listing performance in such situations.

[1]: https://github.com/odoo/odoo/blob/7746621b92c27f6250c0895b0c54b501097e05bb/addons/mrp/models/mrp_bom.py#L160

You can see in this comparison when listing 80 products and having [the `stock_available_mrp` addon](https://odoo-community.org/shop/product/consider-the-production-potential-is-available-to-promise-2400) installed:

Captura de pantalla de 2019-06-07 08-49-59

The time spent in this method is dramatically reduced.

---

TODO

- [ ] When Odoo decides to merge this PR, reopen https://github.com/OCA/stock-logistics-warehouse/pull/611 downstream, which was wrongly merged and it was reverted in https://github.com/OCA/stock-logistics-warehouse/pull/784.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr
@Tecnativa

### CODE REVIEW

**Comment by pedrobaeza** [comment]
What happens when you modify a BoM and it's cached?

**Comment by Yajo** [comment]
Good point, let me clear cache in such situations.

**Comment by Yajo** [comment]
Done, thanks.

File: addons/mrp/models/mrp_bom.py [hunk 1]
```diff
@@ -93,13 +93,25 @@ def onchange_routing_id(self):
     def name_get(self):
         return [(bom.id, '%s%s' % (bom.code and '%s: ' % bom.code or '', bom.product_tmpl_id.display_name)) for bom in self]
 
+    @api.model
+    def create(self, vals):
+        self._bom_find.clear_cache()
```
Comment (pedrobaeza) [inline addons/mrp/models/mrp_bom.py]: You must add a docstring / comment saying why are you doing this. Something like: "Clear cache for getting proper `_bom_find` calls later`

**Comment by pedrobaeza** [comment]
@jco-odoo would you consider this one?

**Comment by Elkasitu** [comment]
IMO it's better to just `groupby` the `bom_lines` by the parameters of `_bom_find` in the body of `explode` and then calling `_bom_find` once per group, `ormcache` can be dangerous if not properly invalidated and in this case it's very local...

That being said, it might be not worth the risk for a stable branch

**Comment by Yajo** [comment]
Please take a quick look here to see how important this can be downstream: https://github.com/OCA/stock-logistics-warehouse/pull/611

I'm not sure how could we do the groupby here... There are a few variables involved.

I think current cache invalidation is pretty good, but we could just invalidate cache at the end of `explode()` if you want, just to be sure. It would still be an improvement on big datasets.

File: addons/mrp/models/mrp_bom.py (hunk 1 above)
Comment (Yajo) [reply addons/mrp/models/mrp_bom.py]: Done.

**Comment by Yajo** [comment]
> `ormcache` can be dangerous if not properly invalidated

I've been reviewing this part because this sentence is very true. However I think it's being properly invalidated in this patch I'm proposing, based on these facts:

- The cache keys `"self._uid", "product_tmpl", "product", "picking_type", "company_id", keys=("company_id",)` mean that in case a different user is calling the method, or in case it's a different product or template product or picking type or company (either explicit or implicit), the cache will be stored differently. So, adding/removing/editing products, templates, picking types, companies or users would never make a failure here.
- The body of the method doesn't use any subfields of any of those arguments, except the ID, which is implicitly stored in the cache.
- The result of the method returns the 1st found BoM that matches the produced criteria, and ordered by sequence or product_id. The only things that could alter that result would be if a BoM is created, edited or deleted. All of those 3 cases invalidate this cache, so it's OK.
- The performance benefit in real world cuts about a 47% of execution/waiting time, as shown in the flamegraph above, so the patch has a high value/risk ratio.
- Tests pass.
- Doesn't violate any guidelines from https://github.com/odoo/odoo/wiki/Contributing#what-does-stable-mean.

The cache is an awesome tool, we shouldn't avoid using it just because of "fear". Just use it carefully, as done here.

**Comment by pedrobaeza** [comment]
One question more: the cache clearing is emptying all the cache or only the cache of this method?

@jco-odoo what do you think of this patch? Should you assign other person for check it?

**Comment by Yajo** [comment]
> the cache clearing is emptying all the cache or only the cache of this method?

AFAIK it should clear it only for that method.

The method I'm calling to do so is added by the decorator itself, so at least that's what I expect it to do: https://github.com/odoo/odoo/blob/7746621b92c27f6250c0895b0c54b501097e05bb/odoo/tools/cache.py#L57

**Comment by pedrobaeza** [comment]
OK, thanks for the clarifications. This seems good to me then, but waiting any feedback.

**Comment by jco-odoo** [comment]
To me, the gain in speed does not compensate for the hacky nature of the patch, although it is indeed interesting to see if we could use the cache more.  To be checked with @sle-odoo and @nim-odoo .

**Comment by rafaelbn** [comment]
@Yajo how easy is to apply this in v12? is it needed? The customer for which we make this is already in v12 migrated :smile: and maybe all this has changed?

**Comment by Yajo** [comment]
> how easy is to apply this in v12? 

Direct.

> is it needed? 

Yes. 

> The customer for which we make this is already in v12 migrated smile and maybe all this has changed?

Does it matter?

---

BTW...

> To me, the gain in speed does not compensate for the hacky nature of the patch

This comment is discouraging, and I don't agree with it. :disappointed: 

This is not hacky at all. I'm just caching a call that is repeated *a lot* of times, even *recursively* with the same parameters.

> although it is indeed interesting to see if we could use the cache more

I do agree here though. It's a powerful tool (as you can see just by looking at the graph).

Why not start today? Hit merge! :yum:

**Comment by rousseldenis** [comment]
@Yajo Wrong rebase I think

**Comment by Yajo** [comment]
Yes, sorry, fixed now.

**Comment by Yajo** [comment]
So, do you want Odoo to be faster or not? :yum:

**Comment by yelizariev** [comment]
Instead of caching we made new method

https://github.com/odoo/odoo/blob/4c62765195a6410da1c4e74c0f1a0fcf57c38e97/addons/mrp/models/mrp_bom.py#L190-L191

Work on ``explode`` method is in progress:
https://github.com/odoo/odoo/pull/68637
https://github.com/odoo-dev/odoo/pull/728

**Comment by Yajo** [comment]
Cool! I'm happy to see this finally getting attention. I'll close as this won't ever be merged anyway. Thanks 😊

**Pull request closed without merging**